eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -286,12 +287,11 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo)
     {
         $dbHandler = $this->getConnection();
         // do not delete assignments if there's other published version containing this field
-        if ($versionInfo !== null &&
-            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+        if ($versionInfo->status === VersionInfo::STATUS_ARCHIVED &&
```

Comment:
> seems you can also pass VersionInfo down to storeFieldData, make it required property and avoid the whole query

Did you mean simplify this line or some db query?

Done like that, please note that for this to work I had to change checking for *archived* version status instead of *not published*. Otherwise tests failed. I guess for drafts I also shouldn't delete keywords as they're used to publish version.
Also, the bug is about deleting archived versions, so after consideration this even makes more sense.